fuel_gauge: Move SBS specific props

aaronemassey opened the pull request:
Moving SBS specific properties to a separate header.

# Immediate followup changes

* Remove the `fuel_gauge_set/get_property` structs in favor of a single property struct.

teburd reviewed (changes requested):
Some of these properties do make sense to remove the SBS prefix from and make them as generic properties (e.g. capacity alarm, time alarm, time to empty, etc)

Removing the SBS specifics though, e.g. SBS MANUFACTURER ACCESS seems like a step in the wrong direction. These integer property mappings aren't in short supply. There will always be device or device class (SBS) specifics that are useful and need to be exposed in some manner.

I'd ask that rather than remove and promise to re-add them, the various sbs_ prefixed properties that can be generalized are, and the sbs specific properties be left alone.

@rgundi @albertofloyd I believe have very specific needs for these properties.

nashif commented:
> Exposing SBS specific properties was a mistaken API choice that lead to leaky abstraction.

care to go into the details and what led to this?

fabiobaltieri commented:
> > Exposing SBS specific properties was a mistaken API choice that lead to leaky abstraction.
> 
> care to go into the details and what led to this?

We were going through this API again now that there's more drivers while reviewing the charger API ones, I suggested the change, that is, moving the SBS specific property definitions in their own file and leave only the generic ones at subsystem level. Reasoning behind it is that every fuel gauge has its own quirks and special device specific properties and it mixing all them up in the same file does not seem like a great idea. This is inline with how the sensor API has device specific extensions (example: https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/sensor/max17055.h), it's used quite extensively and scales pretty well.

fabiobaltieri commented on drivers/fuel_gauge/sbs_gauge/sbs_gauge.c:
```diff
@@ -100,10 +100,6 @@ static int sbs_gauge_get_prop(const struct device *dev, struct fuel_gauge_get_pr
 		rc = sbs_cmd_reg_read(dev, SBS_GAUGE_CMD_AVG_TIME2FULL, &val);
 		prop->value.runtime_to_full = val;
 		break;
-	case FUEL_GAUGE_SBS_MFR_ACCESS:
-		rc = sbs_cmd_reg_read(dev, SBS_GAUGE_CMD_MANUFACTURER_ACCESS, &val);
-		prop->value.sbs_mfr_access_word = val;
-		break;
```
bit confused by the change, are you actually removing these and re-add them later? idea was just to define them in a separate header, no point ripping off the implementation

aaronemassey replied:
I should've made this more explicit in the PR comment. I think working with the SBS properties directly are really the wrong level of abstraction here.

Originally, I was going to just move them to another header as we discussed, but looking at them again, I wasn't sure having SBS specific properties that were more references to the direct registers was a good idea.

It seems like @teburd thinks we definitely need SBS manufacturers access though. Which is fine, we could keep that or define it in an external `sbs.h` so it's not cluttering the existing properties.

aaronemassey commented:
@teburd @nashif @fabiobalteri

Apologies for slow response, got caught up in parallel work.

@teburd, Ack. I'll remove the SBS from the non-SBS specific properties.

Just an FYI, this PR is more of an open question (I should've made that more
clear) and I'm happy to keep SBS manufacturer access. But here is more detail on
my reasoning below:

As far as I understand, the writes/reads for the SBS manufacturers register (as
well as other SBS registers) are often chip specific and are used to accomplish
some other abstract behavior that could be defined at an API level. If this
reasoning is wrong and we want to expose this register directly to folks, I'm
more than happy to leave the manufacturers access as is.

I also think if we want to make that property available as a common
workaround/leaky abstraction for users, I'm also comfortable with that line of
reasoning too.

Happy to receive examples of where we need direct access to SBS manufacturers
access that isn't trying to accomplish a task that can't be defined by the API
as something like: `do_battery_cutoff()`.

teburd commented:
> > > Exposing SBS specific properties was a mistaken API choice that lead to leaky abstraction.
> > 
> > 
> > care to go into the details and what led to this?
> 
> We were going through this API again now that there's more drivers while reviewing the charger API ones, I suggested the change, that is, moving the SBS specific property definitions in their own file and leave only the generic ones at subsystem level. Reasoning behind it is that every fuel gauge has its own quirks and special device specific properties and it mixing all them up in the same file does not seem like a great idea. This is inline with how the sensor API has device specific extensions (example: https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/sensor/max17055.h), it's used quite extensively and scales pretty well.
> 
> The suggestion was just to move the props in a different header though, no need to rip the implementation off and in again.

Ok this makes sense, but I'd say many of the properties are generalizable. E.g. the at rate and time to properties are very likely to be useful at a more general level.

The SBS specifics could be in a battery/sbs.h kind of header available and defined for those that wish to access them still, I like that. We should absolutely keep them though.

aaronemassey commented:
@teburd 

SGTM then. I'll turn this into draft and just move them to an SBS specific header and I'll keep the set function. Thanks for the quick response!

aaronemassey committed:
fuel_gauge: Move SBS API definions

The fuel gauge API header contains definitions that are exceptionally
specific to the SBS spec. While these SBS specific properties are useful to
have, they clutter up the main fuel gauge API and may semantically confuse
clients as to whether the API is SBS specific or not.

Move the SBS properies and definitons to a different header.

Signed-off-by: Aaron Massey <aaronmassey@google.com>

aaronemassey commented on include/zephyr/drivers/fuel_gauge.h:
```diff
@@ -169,18 +152,8 @@ struct fuel_gauge_get_property {
-		uint16_t sbs_remaining_capacity_alarm;
-		/** FUEL_GAUGE_SBS_REMAINING_TIME_ALARM */
-		uint16_t sbs_remaining_time_alarm;
+		/** See SBS spec for value interpretation */
```
I'd rather not just maintain a separate mapping of property to type for something that exists clearly in a spec when we're directly using that spec.

aaronemassey commented on include/zephyr/drivers/fuel_gauge.h:
```diff
@@ -198,16 +171,8 @@ struct fuel_gauge_set_property {
-		/** FUEL_GAUGE_SBS_ATRATE */
-		int16_t sbs_at_rate;
+		/** See SBS spec for value interpretation */
+		uint16_t sbs_word;
```
Adding here. But in a follow-up PR I plan on completely removing the set struct and just keeping a single struct.

aaronemassey commented on include/zephyr/drivers/fuel_gauge_sbs.h:
```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2023 Google LLC
+ * Copyright 2023 Microsoft Corporation
```
Directly copied definitions contributed by Microsoft, so also copied the copyright.

teburd commented on include/zephyr/drivers/fuel_gauge.h:
```diff
@@ -82,25 +78,16 @@ enum fuel_gauge_property {
 	FUEL_GAUGE_DEVICE_CHEMISTRY,
 
+	/** Reserved for start of SBS specific properties. See the fuel_gauge_sbs.h header */
+	_FUEL_GAUGE_SBS_START,
```
Might be nice to follow how sensor.h does this, e.g. 

FUEL_GAUGE_COMMON_COUNT,

FUEL_GAUGE_PRIV_START = FUEL_GAUGE_COMMON_COUNT,

FUEL_GAUGE_MAX = INT16_MAX,

teburd reviewed (commented):
Some minor nits

aaronemassey commented:
Moving this to *draft* mode. @fabiobaltieri Requested I do some of the other refactors in https://github.com/zephyrproject-rtos/zephyr/issues/61818 first.
